ENH: Styler builtins: `highlight_between`

attack68 opened the pull request:
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

- [x] update docs for existing highlight functions for consistency
- [x] add a new `props` argument to existing highlight functions.



Will add tests if approved:
Could wrap both methods into one as well, but probably easier api if two? 

![Screen Shot 2021-02-15 at 17 54 23](https://user-images.githubusercontent.com/24256554/107974441-e44e4d80-6fb6-11eb-8ed0-aaadf0ba2aa2.png)

![Screen Shot 2021-02-15 at 18 20 34](https://user-images.githubusercontent.com/24256554/107976893-90456800-6fba-11eb-93be-4c8b40f8160d.png)

attack68 committed:
merged

attack68 committed:
merged

jreback commented:
I like this, *esp* the smile, have to have it.

jreback commented:
cc @TomAugspurger

attack68 committed:
upodate docs and consistency

attack68 committed:
upodate docs and consistency

attack68 committed:
add and update tests

attack68 committed:
highlight range now accepts sequence

attack68 committed:
highlight range now accepts sequence

attack68 committed:
highlight range now accepts sequence

attack68 committed:
examples

attack68 committed:
reorder

attack68 committed:
Merge remote-tracking branch 'upstream/master' into range_builtin

# Conflicts:
#	pandas/tests/io/formats/test_style.py

attack68 committed:
adj tests

attack68 committed:
adj tests

attack68 committed:
whats new

attack68 committed:
to_numpy()

attack68 committed:
docstring fix

attack68 committed:
test fix

attack68 committed:
improve coding remove conditional

attack68 committed:
docs links

attack68 committed:
Merge remote-tracking branch 'upstream/master' into range_builtin

# Conflicts:
#	pandas/tests/io/formats/style/test_style.py

attack68 committed:
add tests to new modules.

attack68 committed:
add examples and input validation

attack68 committed:
add examples and input validation

attack68 commented:
This is greenish. `test_from_coo` failure unrelated.  Below are new docs screenshots:

![Screen Shot 2021-02-23 at 09 51 11](https://user-images.githubusercontent.com/24256554/108825856-d5255c00-75c3-11eb-9bc0-805af72f5380.png)
![Screen Shot 2021-02-23 at 09 53 25](https://user-images.githubusercontent.com/24256554/108825868-d787b600-75c3-11eb-8270-45e4ee94093a.png)

attack68 committed:
Merge remote-tracking branch 'upstream/master' into range_builtin

attack68 committed:
Merge remote-tracking branch 'upstream/master' into range_builtin

attack68 committed:
Merge remote-tracking branch 'upstream/master' into range_builtin

attack68 committed:
Merge remote-tracking branch 'upstream/master' into range_builtin

attack68 committed:
Merge remote-tracking branch 'upstream/master' into range_builtin

attack68 committed:
Merge remote-tracking branch 'upstream/master' into range_builtin

jreback commented:
lgtm. cc @pandas-dev/pandas-core if anyone has a chance to look / comment on the api additions here.

jreback commented:
also cc @pandas-dev/pandas-triage

jnothman commented:
What's the reasoning behind the naming of parameters start, stop, q_low and
q_high? Behind q_low being exclusive and all others inclusive? Certainly
won't be easy for users to remember.

I know start,stop from integer ranges (as in the python built-in) where
stop indicates an exclusive bound.

Series.between has parameters left, right. Both are inclusive by default
but can be exclusive with a parameter setting. The names aren't great, but
they are there. "highlight_between" might also be a good name.

Alternatives include allowing lte or lt to be specified, and gt,gte
similarly.

jnothman commented:
Example of a possible unified syntax:
series.between(left="10%", right=500, inclusive="left"). Then apply this in
highlight_between. The reduced visibility of the quantile functionality
here might be less than ideal. You could instead have parameters
left_quantile, right_quantile in the between implementation, with either
left or left_quantile to be specified.

attack68 commented:
> What's the reasoning behind the naming of parameters start, stop, 

I looked around for similar named functions. For ranges, start and stop/end seemed pretty common arg inputs.

> q_low and q_high? Behind q_low being exclusive and all others inclusive? 

The exclusive/inclusiveness is derived from `qcut`, nothing I can do about that but document it. For quartiles the `q` input name also seems quite common, but a `q_start` and `q_stop` are alternatives? 

> The names aren't great, but they are there. "highlight_between" might also be a good name. 

Not a fan of `highlight_between`, it feels too detached from things like `date_range`, `range` np.arange` `RangeIndex` etc. 

> Alternatives include allowing lte or lt to be specified, and gt,gte similarly.

Yes this would be nice, but it added complexity to a reasonably lengthy PR: will include if enough support.

jnothman commented:
This is quite unlike arrange, etc unless it is generating objects at equal
interval.

Qcut isn't magical. It seems to estimate the quantiles, then use them as
bin edges to cut. It doesn't even seem appropriate to use here, where
estimating the quantiles and then using `between` to mask would be just as
appropriate.

attack68 committed:
name change: highlight_range to highlight_between

attack68 commented:
- [x] changed name to `highligh_between`
- [x] implemented `inclusive` as arg for boundary with options `True, False, 'right', 'left'`
- [x] added tests for all `ValueError` and the new boundary inclusive options.

@jnothman after some reflection the `range` does feel like it is an object generation, and `between` is more a conditional, so I have made the change (can easily revert one commit). To be honest what swung it was that I always preferred the `left`/`right` arguments over `start`/`stop` and `between` had these.

I think `highlight_quantile` should be a separate function since it has a slightly different signature and there only a few Styler builtin functions so if a user intends to use quantiles as input it should be easily recognisable to use this method. I plan to modify the method to:

- [ ] use `numpy.quantile` instead of `pandas.qcut`
- [ ] pass through the numpy `interpolation` arg for calculating the quantiles.
- [ ] leverage `highlight_between` code passing through the `inclusive` arg with the calculated quantile values for `left` and `right`

Any further suggestions do let me know.

jreback commented:
this is getting too big. can you cut out the non-controverisal, maybe highlite_null into a separate PR

jreback commented on pandas/io/formats/style.py:
```diff
@@ -1635,78 +1637,390 @@ def highlight_null(
+            Left bound for defining the range.
+        right : scalar or datetime-like, or sequence or array-like, default None
+            Right bound for defining the range.
+        inclusive : str or bool, default True
```
look at pd.date_range, inclusie is usualy both, left, right, neither

jnothman replied:
Problem is consistency with Series.between where it is a bool. Perhaps that should be changed first to match date_range in its definition of "inclusive".

attack68 replied:
changed to accept both possibilities. I think Numpy somewhere might use `None` or `'none'` as well. Probably OK for now to have 'some' consistency but not be exactly the same.

`Series.between` does not allow 'left' or 'right', so this might be added as extension. I can add an `good first issue` and see if anyone picks it up

jreback commented on pandas/io/formats/style.py:
```diff
@@ -1635,78 +1637,390 @@ def highlight_null(
+            inclusive: Union[bool, str] = True,
+        ) -> np.ndarray:
+            def realign(x, arg):
+                if np.iterable(x) and not isinstance(x, str):
```
this is not alignment. use .align()

attack68 replied:
align was bad name in code. `reshape` is more representative. added comment.

jreback reviewed (changes requested):
as indicated let's break these changes up (e.g. you added props here?0 let's try to make clear single purpose PRs. have limited time to review, if you keep changing it this will take forever.

so don't add scope, instead break out pieces to simpler / single purpose PRs.

attack68 committed:
split PR

attack68 committed:
split PR

attack68 committed:
split PR

attack68 committed:
split PR

attack68 committed:
split PR

attack68 committed:
split PR

attack68 committed:
split PR

attack68 commented:
> as indicated let's break these changes up (e.g. you added props here?0 let's try to make clear single purpose PRs. have limited time to review, if you keep changing it this will take forever.
> 
> so don't add scope, instead break out pieces to simpler / single purpose PRs.

@jreback sorry, underestimated the scope at start and then it slowly grew. Appreciate you have a lot to go through. I have split this into 3 PRs:  #40242, this one, and #40244 (wip)

attack68 committed:
Merge remote-tracking branch 'upstream/master' into range_builtin

attack68 committed:
Merge remote-tracking branch 'upstream/master' into range_builtin

attack68 committed:
remove reference to highlight_quantile

attack68 committed:
Merge remote-tracking branch 'upstream/master' into range_builtin

# Conflicts:
#	doc/source/reference/style.rst
#	doc/source/whatsnew/v1.3.0.rst
#	pandas/tests/io/formats/style/test_highlight.py

attack68 committed:
link see alsos

attack68 committed:
doc edit

attack68 committed:
doc edit

attack68 committed:
Merge branch 'master' into range_builtin

jreback commented on pandas/io/formats/style.py:
```diff
@@ -1796,6 +1800,163 @@ def f(data: FrameOrSeries, props: str) -> np.ndarray:
+            Left bound for defining the range.
+        right : scalar or datetime-like, or sequence or array-like, default None
+            Right bound for defining the range.
+        inclusive : {'both', 'neither', 'left', 'right'} or bool, default True
```
can we not include bool and just have a sensisble default? both?

attack68 replied:
of course, I included `bool` for consistency with `Series.between` which only accepts `bool`, and since I thought it was a natural expectation to the `inclusive?` argument. Its inclusion didn't seem to big factor in code.

changed.

attack68 committed:
remove bool option from inclusive

jnothman commented on pandas/io/formats/style.py:
```diff
@@ -1796,6 +1800,163 @@ def f(data: FrameOrSeries, props: str) -> np.ndarray:
+
+            # get ops with correct boundary attribution
+            if inclusive == "both":
+                ops = ("__ge__", "__le__")
```
that is:

```suggestion
                ops = (operator.ge, operator.le)
```

attack68 replied:
changed. thx for tip never knew that.

attack68 committed:
Merge branch 'master' into range_builtin

attack68 committed:
use operator

jnothman commented on pandas/tests/io/formats/style/test_highlight.py:
```diff
@@ -63,3 +63,65 @@ def test_highlight_minmax_ext(self, f, kwargs):
+
+    @pytest.mark.parametrize(
+        "kwargs",
+        [
```
I'm confused about how these all pass when there is self.randn(2) in self.df, which often will be outside of [0, 1]. Maybe I've missed something.

attack68 replied:
the random seed is set so self.df is always: [[0, -0.6], [1, -1.2]].
where negatives are included in the potential highlighting range, the `subset` argument is used to restrict application only to col "A".

attack68 replied:
incidentally I don't know why random is used. just complicates things, but is legacy from my perspective.

jnothman replied:
Yes, I'm more used to other conventions around randomisation in testing. (As far as I'm concerned, if there is randomisation on the global generator, the test should be presumed invariant to the initial random state.)

Given that the random df is not explicitly shown near this test, could you please include a comment explaining, therefore, why these parameters make sense: even just show the df, or set it explic itly.

attack68 replied:
i have set the df explicitly keeping the values same as previous. all tests still pass, of course.

attack68 committed:
remove misleading random note from test fixtures

attack68 committed:
remove misleading random note from test fixtures

attack68 committed:
Merge remote-tracking branch 'upstream/master' into range_builtin

attack68 committed:
mypy fix

jnothman commented on pandas/io/formats/style.py:
```diff
@@ -1808,6 +1813,163 @@ def f(data: FrameOrSeries, props: str) -> np.ndarray:
+        ... })
+        >>> df.style.highlight_between(left=2.1, right=2.9)
+
+        .. figure:: ../../_static/style/hr_basic.png
```
Might want to change the names of these files.

jnothman reviewed (approved):
Otherwise LGTM

attack68 committed:
Merge remote-tracking branch 'upstream/master' into range_builtin

attack68 committed:
filesnames

attack68 committed:
Merge remote-tracking branch 'upstream/master' into range_builtin

attack68 committed:
add arg shape validation function

attack68 commented on pandas/io/formats/style.py:
```diff
@@ -2257,3 +2414,59 @@ def pred(part) -> bool:
     return tuple(slice_)
+
+
+def _validate_apply_axis_arg(
```
this arg validation was added to remove code duplication with #39930 and it might be also be useful for any other dev

attack68 committed:
Merge remote-tracking branch 'upstream/master' into range_builtin

attack68 committed:
mypy updates

attack68 committed:
Merge remote-tracking branch 'upstream/master' into range_builtin

attack68 committed:
Merge remote-tracking branch 'upstream/master' into range_builtin

attack68 committed:
Merge remote-tracking branch 'upstream/master' into range_builtin

attack68 committed:
Merge remote-tracking branch 'upstream/master' into range_builtin

# Conflicts:
#	pandas/tests/io/formats/style/test_highlight.py

attack68 committed:
integrate into new test structure

attack68 committed:
Merge remote-tracking branch 'upstream/master' into range_builtin

attack68 committed:
remove axis from validate arg func

attack68 committed:
Merge remote-tracking branch 'upstream/master' into range_builtin

# Conflicts:
#	pandas/io/formats/style.py

attack68 committed:
no bool in inclusive

attack68 committed:
revert random explicit

jreback commented:
can you merge master

attack68 committed:
Merge remote-tracking branch 'upstream/master' into range_builtin

attack68 commented:
@jreback ping on merge and green

jreback commented:
thanks @attack68 very nice.

can you make sure that the notebook is up to date after last few things merged.

attack68 commented:
> thanks @attack68 very nice.
> 
> can you make sure that the notebook is up to date after last few things merged.

yes I will do that in coming week.
